File: p275m/h_index2.js
/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function (citations) {
  const N = citations.length;
  let [lo, hi] = [0, N]; // min and max of hIndex
  while (lo < hi) {
    let mid = ~~((lo + hi) / 2) + 1; // go ceiling
    if (citations[N - mid] >= mid) {
      lo = mid;
    } else {
      hi = mid - 1;
    }
  }
  return lo;
};
// TEST
[
  [[8, 8, 8, 8, 9], 5],
  [[0, 0, 0, 0, 0, 0], 0],
  [[0, 0, 0, 0, 0, 100], 1],
  [[0, 1, 3, 5, 6], 3],
  [[0, 1, 2, 2, 3, 4, 8], 3],
].forEach((t) => {
  let act = hIndex(t[0]);
  console.log('H-Index of', t[0], '->', act);
  console.assert(act === t[1]);
});

File: p275m/h_index2.py
from typing import List


class Solution:
    def hIndex(self, citations: List[int]) -> int:
        N = len(citations)
        l, h = 0, N
        while l < h:
            m = (l + h) // 2 + 1
            if citations[N - m] >= m:
                l = m
            else:
                h = m - 1
        return l


# TESTS
tests = [
    [[8, 8, 8, 8, 9], 5],
    [[0, 0, 0, 0, 0, 0], 0],
    [[0, 0, 0, 0, 0, 100], 1],
    [[0, 1, 3, 5, 6], 3],
    [[0, 1, 2, 2, 3, 4, 8], 3],
]
for t in tests:
    sol = Solution()
    actual = sol.hIndex(t[0])
    print("H-Index of", t[0], "->", actual)
    assert actual == t[1]
